### PR TITLE
[SDK] marketplace v3 integration

### PR DESCRIPTION
**Pull request by kumaryash90**

### CODE REVIEW

File: packages/sdk/src/evm/core/classes/marketplacev3-direct-listings.ts [hunk 1]
```diff
@@ -129,19 +142,33 @@ export class MarketplaceV3DirectListings {
...
       );
+      start += DEFAULT_QUERY_ALL_COUNT;
+    }
+    partialListings.push(
+      await this.contractWrapper.readContract.getAllValidListings(
+        start,
+        end - 1,
+      ),
+    );
+    rawListings = (await Promise.all(partialListings)).flat();
```
Comment (joaquim-verges) [inline packages/sdk/src/evm/core/classes/marketplacev3-direct-listings.ts:167]: this code is duplicated in a few places - can we make a helper function for it? there's already a common/marketplace.ts file where you can put this

**Comment by nachoiacovino** [comment]
Version from this release is: [3.7.0-dev-4155fe3](https://www.npmjs.com/package/@thirdweb-dev/sdk/v/3.7.0-dev-4155fe3)

File: packages/sdk/src/evm/common/marketplacev3.ts [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+ *
+ * @internal
+ * @param provider - The connected provider
+ * @param transferrerContractAddress - The address of the marketplace contract
+ * @param assetContract - The address of the asset contract.
+ * @param tokenId - The token id of the token.
+ * @param owner - The address of the account that owns the token.
+ * @returns - True if the transferrerContractAddress is approved on the token, false otherwise.
+ */
+export async function isTokenApprovedForTransfer(
```
Comment (joaquim-verges) [inline packages/sdk/src/evm/common/marketplacev3.ts:25]: these should not be duplicated

File: packages/sdk/src/evm/contracts/prebuilt-implementations/marketplacev3.ts [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+      this.storage,
+    );
+    this.roles = new ContractRoles(
+      this.contractWrapper,
+      MarketplaceV3.contractRoles,
+    );
+    this.encoder = new ContractEncoder(this.contractWrapper);
+    this.estimator = new GasCostEstimator(this.contractWrapper);
+    this.directListings = new MarketplaceV3DirectListings(
+      this.contractWrapper as unknown as ContractWrapper<DirectListingsLogic>,
```
Comment (joaquim-verges) [inline packages/sdk/src/evm/contracts/prebuilt-implementations/marketplacev3.ts:213]: re-add feature detection here pls

File: packages/sdk/src/evm/core/classes/marketplacev3-direct-listings.ts [hunk 4]
```diff
@@ -0,0 +1,833 @@
...
+import { BigNumber, BigNumberish, Contract } from "ethers";
+import invariant from "tiny-invariant";
+
+/**
+ * Handles direct listings
+ * @public
+ */
+export class MarketplaceV3DirectListings {
+  private contractWrapper: ContractWrapper<DirectListingsLogic>;
+  private storage: ThirdwebStorage;
```
Comment (joaquim-verges) [inline packages/sdk/src/evm/core/classes/marketplacev3-direct-listings.ts:48]: can you add ContractEvents / ContractEncoder and ContractInterceptor to the plugins as well?

**Comment by nachoiacovino** [comment]
wonder if we should prefix the extensions to be `MarketplaceOffers`, `MarketplaceEnglishAuctions`, etc.

**Pull request merged**